contracts/SwapKiwi.sol

```diff
@@ -65,131 +74,166 @@ contract SwapKiwi is Ownable, IERC721Receiver {
-        _swaps[swapId].secondUserEtherValue = msg.value - fee;
-        _etherLocked += _swaps[swapId].secondUserEtherValue;
-      }
-
-      emit SwapInitiated(
-        msg.sender,
-        _swaps[swapId].initiator,
-        swapId,
-        nftAddresses,
-        nftIds,
-        _swaps[swapId].secondUserEtherValue
-      );
+    _swaps[swapId].secondUserNftAddresses = nftAddresses;
+    _swaps[swapId].secondUserNftIds = nftIds;
+    _swaps[swapId].secondUserNftAmounts = nftAmounts;
+
+    uint128 _fee = fee;
+
+    if (msg.value > _fee) {
+      _swaps[swapId].secondUserEtherValue = uint128(msg.value) - _fee;
+      _etherLocked += _swaps[swapId].secondUserEtherValue;
+    }
+
+    emit SwapInitiated(
+      msg.sender,
+      _swaps[swapId].initiator,
+      swapId,
+      _swaps[swapId].secondUserEtherValue,
+      nftAddresses,
+      nftIds,
+      nftAmounts
+    );
   }
 
   /**
-    * @dev Swap initiator accepts the swap (NFTs proposed by the second user).
-    *      Executeds the swap - transfers NFTs from SwapKiwi to the participating users.
-    *      Callable only by swap initiator.
-    *
-    * @param swapId ID of the swap that the initator wants to execute
-    */
-  function acceptSwap(uint256 swapId) external onlyInitiator(swapId) {
+  * @dev Swap initiator accepts the swap (NFTs proposed by the second user).
+  *      Executeds the swap - transfers NFTs from SwapKiwi to the participating users.
+  *      Callable only by swap initiator.
+  *
+  * @param swapId ID of the swap that the initator wants to execute
+  */
+  function acceptSwap(uint64 swapId) external onlyInitiator(swapId) {
     require(
       (_swaps[swapId].secondUserNftAddresses.length != 0 || _swaps[swapId].secondUserEtherValue > 0) &&
       (_swaps[swapId].initiatorNftAddresses.length != 0 || _swaps[swapId].initiatorEtherValue > 0),
-       "SwapKiwi: Can't accept swap, both participants didn't add NFTs"
+      "SwapKiwi: Can't accept swap, both participants didn't add NFTs"
     );
 
     // transfer NFTs from escrow to initiator
     safeMultipleTransfersFrom(
       address(this),
       _swaps[swapId].initiator,
       _swaps[swapId].secondUserNftAddresses,
-      _swaps[swapId].secondUserNftIds
+      _swaps[swapId].secondUserNftIds,
+      _swaps[swapId].secondUserNftAmounts
     );
 
     // transfer NFTs from escrow to second user
     safeMultipleTransfersFrom(
       address(this),
       _swaps[swapId].secondUser,
       _swaps[swapId].initiatorNftAddresses,
-      _swaps[swapId].initiatorNftIds
+      _swaps[swapId].initiatorNftIds,
+      _swaps[swapId].initiatorNftAmounts
     );
 
     if (_swaps[swapId].initiatorEtherValue != 0) {
       _etherLocked -= _swaps[swapId].initiatorEtherValue;
-      uint amountToTransfer = _swaps[swapId].initiatorEtherValue;
+      uint128 amountToTransfer = _swaps[swapId].initiatorEtherValue;
       _swaps[swapId].initiatorEtherValue = 0;
       _swaps[swapId].secondUser.transfer(amountToTransfer);
```

Comment:
Due to possible future gas cost changes, it is discouraged to use address.transfer, instead should use address.call with success verification.